### PR TITLE
[ADD] Add module base_conditional_image

### PR DESCRIPTION
**Pull request by p-tombez**
Addition of a custom image module.
As soon as it is accepted, I'll port it to V10+

### CODE REVIEW

**Review by tschanzt — APPROVED** [review]
code review only. but looks good.

File: custom_image/models/image.py
Comment (simahawk) [inline custom_image/models/image.py]: distinguish images for what?

File: custom_image/models/abstract_image.py
Comment (simahawk) [inline custom_image/models/abstract_image.py]: self._name is enough

File: custom_image/models/abstract_image.py
Comment (simahawk) [inline custom_image/models/abstract_image.py]: ?

File: custom_image/models/abstract_image.py
Comment (simahawk) [inline custom_image/models/abstract_image.py]:
```python
if 'company_id' in self:
```
or, to be more explicit
```python
if 'company_id' in self._fields:
```

File: custom_image/models/abstract_image.py
Comment (simahawk) [inline custom_image/models/abstract_image.py]: you are already searching, do this check in the search above instead

File: custom_image/README.rst
Comment (simahawk) [inline custom_image/README.rst]: to achieve what?
I'll put review on hold until you clarify what are the main goals and - for instance - how is this different from base_multi_image and/or if it's compatible w/ it :wink: 

A real use case might help

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by p-tombez** [comment]
I fixed the code according to the comments and also rephrased the module description as well as renaming it to better match its actual purpose

File: base_conditional_image/models/abstract_image.py
Comment (acsonefho) [inline base_conditional_image/models/abstract_image.py]: `cls` is not defined, you can use `self`.

File: base_conditional_image/models/abstract_image.py
Comment (acsonefho) [inline base_conditional_image/models/abstract_image.py]: It's quite hard to read/understand, can you rewrite it please?

File: base_conditional_image/models/image.py
Comment (acsonefho) [inline base_conditional_image/models/image.py:48]: This function should be `api.model`

File: base_conditional_image/views/image_view.xml
Comment (acsonefho) [inline base_conditional_image/views/image_view.xml:3]: Entire file is bad intended (should be 4 spaces)

File: base_conditional_image/views/image_view.xml
Comment (acsonefho) [inline base_conditional_image/views/image_view.xml:2]: I think it should be `odoo` and not `openerp`

File: base_conditional_image/README.rst
Comment (acsonefho) [inline base_conditional_image/README.rst]: This file should be generated automatically.
You have to create a readme directory with every sections of the `README.rst`.

**Review by acsonefho — CHANGES_REQUESTED** [review]

File: base_conditional_image/views/image_view.xml
Comment (leemannd) [reply base_conditional_image/views/image_view.xml:3]: @acsonefho AFAIR, there is no common agreement on the tab space into OCA repos.

File: base_conditional_image/views/image_view.xml
Comment (p-tombez) [reply base_conditional_image/views/image_view.xml]: It's a V9 branch, so it's ```openerp``` not ```odoo```

Comment (p-tombez) [reply base_conditional_image/views/image_view.xml:2]: It's a V9 branch, so it's ```openerp``` not ```odoo```

File: base_conditional_image/models/abstract_image.py
Comment (p-tombez) [reply base_conditional_image/models/abstract_image.py]: You're right, I missed it during refactoring

File: base_conditional_image/models/abstract_image.py
Comment (p-tombez) [reply base_conditional_image/models/abstract_image.py]: I don't really see how to rewrite that, do you have a proposal ?

Comment (acsonefho) [reply base_conditional_image/models/abstract_image.py]: No I don't, sorry =/ But I think it could be more readable maybe with 2 `if`?

File: base_conditional_image/views/image_view.xml
Comment (acsonefho) [reply base_conditional_image/views/image_view.xml:3]: Ok ok no problem, I thought there was

File: base_conditional_image/views/image_view.xml
Comment (acsonefho) [reply base_conditional_image/views/image_view.xml:2]: Ok ok, I think I work too much on v10.0 :)

File: base_conditional_image/README.rst
Comment (p-tombez) [reply base_conditional_image/README.rst]: It's a V9 branch, no module are using this pattern. I'll switch to it when I migrate the module to V10 and later

Comment (acsonefho) [reply base_conditional_image/README.rst]: I think it's not related to the version.
No module in v9.0 using this pattern because this pattern is quite recent

Comment (leemannd) [reply base_conditional_image/README.rst]: AFAIK it's a recommendation, and it's not mandatory. And the readme is really explicit.
@acsonefho Do you think your comment on this point can be dismissed?

Comment (acsonefho) [reply base_conditional_image/README.rst]: @leemannd Some say that's required and others just a recommendation.
On the documentation (on maintainer-tools repo), the readme file is mandatory and the way to generate it is with the `readme` directory and some files.
I think that someone should take a final decision about that.
I'll not "block" the MR for this of course!

File: base_conditional_image/__openerp__.py
Comment (simahawk) [inline base_conditional_image/__openerp__.py]: you know... :)

File: base_conditional_image/__openerp__.py
Comment (simahawk) [inline base_conditional_image/__openerp__.py]: c2c 1st, OCA always at last place

File: base_conditional_image/README.rst
Comment (simahawk) [reply base_conditional_image/README.rst]: Usually is not considered mandatory but for NEW modules I'd argue it would be :man_teacher: 
Plus, splitting this to readme folder is a 5mins operation :wink:

File: base_conditional_image/models/abstract_image.py
Comment (simahawk) [reply base_conditional_image/models/abstract_image.py]: I'd do it like this more or less:
```python
    [...]

    @staticmethod
    def _compute_image_test_without_company(image, rec):
        return bool(safe_eval(image.selector or "True", {'object': rec})))

    @staticmethod
    def _compute_image_test_with_company(image, rec):
        if (image.company_id == rec.company_id or 
                rec.company_id and not image.company_id):
            return self._compute_image_test_without_company(image, rec)
        return False

    def _compute_image(self):
        domain = [('model_name', '=', self._name)]
        if 'company_id' in self._fields:
            test_func = self._compute_image_test_with_company
        else:
            # If inherited object doesn't have a `company_id` field,
            # remove the items with a company defined and the related checks
            domain.append(('company_id', '=', False))
            test_func = self._compute_image_test_without_company

        images = self.env['image'].search(domain, order='company_id, selector')
        for rec in self:
            found = None
            for image in images:
                if test_func(image, rec):
                    found = image
                    break
            if found:
                rec.update({
                    'image': found.image,
                    'image_medium': found.image_medium,
                    'image_small': found.image_small,
                })
```

**Review by simahawk — CHANGES_REQUESTED** [review]

File: base_conditional_image/models/abstract_image.py
Comment (simahawk) [inline base_conditional_image/models/abstract_image.py]: @p-tombez this thing can be reaaaaally expensive. Are we sure there's no way to limit this?

Comment (p-tombez) [reply base_conditional_image/models/abstract_image.py]: I don't see what could be expensive here nor any way of limiting anything more. There is already an exit condition as soon as a result is found...

Comment (simahawk) [reply base_conditional_image/models/abstract_image.py]: @p-tombez an I wrong or potentially you can have gazillion of images here?

Comment (p-tombez) [reply base_conditional_image/models/abstract_image.py]: I think you're wrong ;)
The recordset ```images``` will only hold the image "selectors" for the current object.
Indeed, the size will depends on how the user configured these selectors, but I think it will not go over a dozen

**Review by guewen — APPROVED** [review]

File: base_conditional_image/models/abstract_image.py
Comment (simahawk) [reply base_conditional_image/models/abstract_image.py]: @p-tombez ok, then pls add a comment of top of it saying that it should be fine in most of the cases because bla bla bla ;)

**Comment by simahawk** [comment]
@p-tombez build is red. Also, pls, add the comment mentioned in https://github.com/OCA/server-tools/pull/1377#discussion_r248267223

**Comment by p-tombez** [comment]
@simahawk In my previous commit (https://github.com/OCA/server-tools/pull/1377/commits/7c7bcf46b9741479be2d34e74cc3b0636e1cff1d) I changed the names to avoid confusion. I think the code is now pretty self-explanatory and does not require a comment anymore (unless you don't know how nested for loop works).
For the red build, I will not fix the lint issues in the other modules (at least, not in this PR) and the errors in the module ```oauth_provider``` were already there before I made this PR (https://travis-ci.org/OCA/server-tools/builds/435521532?utm_source=github_status&utm_medium=notification)

**Comment by simahawk** [comment]
@p-tombez fine, naming works better. Build is still red tho.

**Pull request closed without merging**
